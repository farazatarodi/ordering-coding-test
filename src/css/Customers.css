.customers-container {
  width: 80%;
  height: fit-content;
}

.customers-content {
  margin: 3rem 0;
  display: flex;
  flex-wrap: wrap;
  width: 100%;
  column-gap: 2%;
  align-self: center;
  border: solid 2px var(--color-neutral);
  box-shadow: var(--shadow);
  border-radius: var(--radius);
}

.customers-header {
  width: 100%;
  display: grid;
  grid-template-columns: 10% auto 20% 20%;
  padding: 2rem;
  background-color: var(--color-neutral-light);
  border-radius: var(--radius) var(--radius) 0 0;
}
