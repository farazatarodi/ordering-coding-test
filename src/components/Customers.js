// main component of customers page

// Redux imports
import { useSelector } from 'react-redux';

// component imports
import CustomerCard from './CustomerCard';
import EmptyCard from './EmptyCard';
import LoadingCard from './LoadingCard';

// css imports
import '../css/Customers.css';
import ErrorCard from './ErrorCard';

const Customers = () => {
  // get customers data from store
  const customersState = useSelector((state) => state.customersReducer);
  const customers = customersState.customers;
  const loading = customersState.loading;
  const error = customersState.error;

  return (
    <div className="customers-container">
      <h1 className="heading heading-1">Customers</h1>
      <div className="customers-content">
        <div className="customers-header heading heading-2">
          <strong>ID</strong>
          <strong>Name</strong>
          <strong>Since</strong>
          <strong>Revenue</strong>
        </div>
        {/* map customers data to individual card components */}
        {loading ? ( // check if loading.
          <LoadingCard /> // render loading card if loading.
        ) : error ? ( // check if fetching has error.
          <ErrorCard error={error} /> // render error card if error.
        ) : customers && customers.length !== 0 ? ( // check if we have customers and it is not empty.
          customers.map((customer) => (
            <CustomerCard
              key={customer.id}
              id={customer.id}
              name={customer.name}
              since={customer.since}
              revenue={customer.revenue}
            />
          ))
        ) : (
          <EmptyCard /> // render empty card if no customers
        )}
      </div>
    </div>
  );
};

export default Customers;
