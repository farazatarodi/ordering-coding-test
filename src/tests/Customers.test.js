import { render, screen } from '@testing-library/react';

import Customers from '../components/Customers';

import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

test('we see the empty table in customers page', () => {
  const initialState = {};
  const mockStore = configureStore();

  const store = mockStore(initialState);
  render(
    <Provider store={store}>
      <Customers />
    </Provider>
  );
  expect(screen.getByText(/id/i)).toBeTruthy();
  expect(screen.queryByText(/â‚¬/i)).toBeNull();
});

test('customer card shows all info correctly', () => {
  const initialState = {
    customersReducer: [
      {
        id: 'testID',
        name: 'testName',
        since: 'testSince',
        revenue: 'testRevenue',
      },
    ],
  };
  const mockStore = configureStore();

  const store = mockStore(initialState);
  render(
    <Provider store={store}>
      <Customers />
    </Provider>
  );
  expect(screen.getByText(/testid/i)).toBeTruthy();
  expect(screen.getByText(/testname/i)).toBeTruthy();
  expect(screen.getByText(/testsince/i)).toBeTruthy();
  expect(screen.getByText(/testrevenue/i)).toBeTruthy();
});
