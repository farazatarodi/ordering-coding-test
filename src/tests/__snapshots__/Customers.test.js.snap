// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correclty 1`] = `
<div
  className="customers-container"
>
  <h1
    className="heading heading-1"
  >
    Customers
  </h1>
  <div
    className="customers-content"
  >
    <div
      className="customers-header heading heading-2"
    >
      <strong>
        ID
      </strong>
      <strong>
        Name
      </strong>
      <strong>
        Since
      </strong>
      <strong>
        Revenue
      </strong>
    </div>
    <div
      className="card text text-display notification-card"
    >
      Nothing to show
    </div>
  </div>
</div>
`;
